lib/runtime-core/src/parse.rs

```diff
@@ -266,28 +266,22 @@ pub fn read_module<
                                 body_begun = true;
                                 fcg.begin_body(&info.read().unwrap())
                                     .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
-                                middlewares
-                                    .run(
-                                        Some(fcg),
-                                        Event::Internal(InternalEvent::FunctionBegin(id as u32)),
-                                        &info.read().unwrap(),
-                                    )
-                                    .map_err(|x| LoadError::Codegen(x))?;
+                                events
+                                    .push(Event::Internal(InternalEvent::FunctionBegin(id as u32)));
                             }
-                            middlewares
-                                .run(Some(fcg), Event::Wasm(op), &info.read().unwrap())
-                                .map_err(|x| LoadError::Codegen(x))?;
+                            events.push(Event::WasmOwned(op.clone()));
+                        }
+                        ParserState::EndFunctionBody => {
+                            events.push(Event::Internal(InternalEvent::FunctionEnd));
```

Comment:
I think we can keep the old style where we call middlewares at each step to prevent filling a buffer with the wasm ops.  That way middleware can do its own lookahead.  The one structural change we'll need to make is ensuring that middleware and our code can handle being given nothing or given more at once.

I'm still fond of using iterators, something like https://github.com/wasmerio/wasmer/pull/1021 , but I ran into issues with all the traits the last time I touched that and due to wasm op not being clone, it wasn't possible

Reply:
>  That way middleware can do its own lookahead. 

I am still unsure how to achieve this with the current streaming design of WasmParser. 

> I'm still fond of using iterators, something like #1021 

How would this allow for a middleware to inject code? How would that affect the iterator? Perhaps this is a rudimentary Rust question, in which case just let me know and I'll do the appropriate Googling.

Reply:
The way middleware would work is by allowing any link in the chain to hold on to things.

So for example given the chain:

```
Parser -> A -> B -> C -> CodeGen
```

Any step can hold on to as many events as it likes.  So if `A` needs to have full lookahead then it would read from Parser until parser stopped returning anything and then start returning to `B`.

If `A` needs to look at functions and `B` needs to look at loops, then `A` can track the state of the streaming opcodes, when it sees a `function start` event it caches events until it sees `function end` and then dumps everything (including any modifications it made into B).  B does the same with loops and so on.

The Rust Iterator trait defines a function `next(&mut self) -> Option<Self::Item>` with the logic being that it returns `Some(T)` while it has items to return and returns `None` when there's nothing left to return and the iterator is done.

Using iterators, we'd have the `CodeGen` step get an opaque iterator and call `next` on it doing whatever it wants with it to generate code.

That iterator is step `C`, what `C` does is use its internal iterator to get events and then do logic on them, so when `C`.next() is called, `B`.next() is called which calls `A`.next() which calls `Parser`.next().  At any step in this chain, an iterator can call `next` on its internal iterator as many times as it wants allowing it to lookahead as much as it wants.  To be more specific about how it'll commonly work: once it's read all that it wants to into its internal buffer, the struct implementing it sets a flag or otherwise tracks that it's ready to return and returns the first item.  When that bit is set, it doesn't call `next()`, it just dumps its internal buffer/outputs its own edits and additions to the stream.

I like this API design a lot because it's very elegant, none of the links in the chain need to know about anything other than how to get more input.

Reply:
And because Iterator is well supported in Rust, if you wanted to just get everything all at once, that's very easy to do, you can just do `internal_iter.collect::<Vec<_>>()` and you have a vector of everything.  In the future when generators and `yield` are in stable Rust, this type of API can be even nicer.  Instead of making an iterator explicitly, you just write a normal function but use the keyword `yield` instead of `return` for each element and magic happens that tracks the state of execution in your function for you.

Reply:
Sounds like this is how it should work. I am not sure how to implement it but if you can then I will modify the middleware to use that approach.